source/common/http/alternate_protocols_cache_impl.h

```diff
@@ -20,37 +20,52 @@ namespace Http {
 
 // An implementation of AlternateProtocolsCache.
 // See: source/docs/http3_upstream.md
+//
+// The primary purpose of this cache is to cache alternate protocols entries.
+// Secondarily, it maps origins to rtt and bandwidth information, useful for
+// tuning 0-rtt timeouts if the alternate protocol is HTTP/3.
 class AlternateProtocolsCacheImpl : public AlternateProtocolsCache,
                                     Logger::Loggable<Logger::Id::alternate_protocols_cache> {
 public:
   AlternateProtocolsCacheImpl(TimeSource& time_source, std::unique_ptr<KeyValueStore>&& store,
                               size_t max_entries);
   ~AlternateProtocolsCacheImpl() override;
 
+  // Captures the data tracked per origin; the alternate protocols supported,
+  // and last smoothed round trip time and bandwidth, if available.
+  struct OriginData {
+    std::vector<AlternateProtocol> protocols;
+    int64_t srtt;
+    int64_t bandwidth;
+  };
+
   // Converts an Origin to a string which can be parsed by stringToOrigin.
   static std::string originToString(const AlternateProtocolsCache::Origin& origin);
   // Converts a string from originToString back to structured format.
   static absl::optional<AlternateProtocolsCache::Origin> stringToOrigin(const std::string& str);
 
-  // Convert an AlternateProtocol vector to a string to cache to the key value
+  // Convert origin data to a string to cache to the key value
   // store. Note that in order to determine the lifetime of entries, this
   // function will serialize ma= as absolute time from the epoch rather than
   // relative time.
   // This function also does not do standards-required normalization. Entries requiring
   // normalization will simply not be read from cache.
-  static std::string protocolsToStringForCache(const std::vector<AlternateProtocol>& protocols,
-                                               TimeSource& time_source);
-  // Parse an alternate protocols string into structured data, or absl::nullopt
+  // The string format is
+  // srtt|bandwidth|protocols
+  static std::string originDataToStringForCache(const std::vector<AlternateProtocol>& protocols,
+                                                int64_t srtt, int64_t bandwidth);
+  // Parse an origin data into structured data, or absl::nullopt
   // if it is empty or invalid.
   // If from_cache is true, it is assumed the string was serialized using
   // protocolsToStringForCache and the the ma fields will be parsed as absolute times
   // rather than relative time.
-  static absl::optional<std::vector<AlternateProtocol>>
-  protocolsFromString(absl::string_view protocols, TimeSource& time_source,
-                      bool from_cache = false);
+  static absl::optional<OriginData> originDataFromString(absl::string_view protocols,
```

Comment:
nit: since the string argument is not longer just the protocols maybe we should name it "data"?

Reply:
done